lib/utils/ignore.ts

```diff
@@ -295,44 +295,40 @@ export async function filterFilesWithDockerignore(
 		? Object.keys(ignoreByDir).filter((dir) => dir && dir !== root)
 		: [];
 	const dockerignoreFiles: FileStats[] = [];
-	const filteredFileList: FileStats[] = [];
-	const klaw = await import('klaw');
-
-	await new Promise((resolve, reject) => {
-		// Looking at klaw's source code, `preserveSymlinks` appears to only
-		// afect the `stats` argument to the `data` event handler
-		klaw(projectDir, { preserveSymlinks: false })
-			.on('error', reject)
-			.on('end', resolve)
-			.on('data', (item: { path: string; stats: Stats }) => {
-				const { path: filePath, stats } = item;
-				// With `preserveSymlinks: false`, filePath cannot be a symlink.
-				// filePath may be a directory or a regular or special file
-				if (!stats.isFile()) {
-					return;
-				}
-				const relPath = path.relative(projectDir, filePath);
-				const fileInfo = {
-					filePath,
-					relPath,
-					stats,
-				};
-				if (path.basename(relPath) === '.dockerignore') {
-					dockerignoreFiles.push(fileInfo);
-				}
-				for (const dir of dockerignoreServiceDirs) {
-					if (relPath.startsWith(dir)) {
-						if (!ignoreByDir[dir].ignores(relPath.substring(dir.length))) {
-							filteredFileList.push(fileInfo);
-						}
-						return;
-					}
-				}
-				if (!ignoreByDir[root].ignores(relPath)) {
-					filteredFileList.push(fileInfo);
-				}
-			});
+
+	const rfs = await import('recursive-fs');
+	const { files } = await rfs.read(projectDir);
+
+	const filteredFilePaths: string[] = (files as string[]).filter((filePath) => {
+		const relPath = path.relative(projectDir, filePath);
```

Comment:
```suggestion
		const relPath = path.relative(projectDir, filePath);
		if (path.basename(relPath) === '.dockerignore') {
			return true;
		}
```
I think (I didn't test) this change is needed to preserve existing behaviour in some corner cases, _in addition to_ the similar lines later in this same function:
```ts
		if (path.basename(relPath) === '.dockerignore') {
			dockerignoreFiles.push(fileInfo);
		}
```
Because the latter lines would be pointless if `.dockerignore` was not included in the `filteredFilePaths` array.

This may (or not) indeed be one of the reasons why some test cases are still failing, e.g.:

```
      -  "[Warn] * C:\\var\\lib\\concourse\\worker\\volumes\\live\\0ac9c696-c6d6-4eba-5984-b6b9c8547f60\\volume\\tests\\test-data\\projects\\no-docker-compose\\dockerignore2/lib/.dockerignore"
      +  "[Warn] * C:\\var\\lib\\concourse\\worker\\volumes\\live\\0ac9c696-c6d6-4eba-5984-b6b9c8547f60\\volume\\tests\\test-data\\projects\\no-docker-compose\\dockerignore2\\lib\\.dockerignore"
```
https://ci.product-os.io/builds/42293

Reply:
Oh, actually, I should read my own comments, :-) the diff I pasted above shows that the difference is related to forward slashes vs backslashes, the latter the canonical form on Windows. It looks like the `recursive-fs` implementation assumes forward slashes on all platforms:

https://github.com/simov/recursive-fs/blob/master/lib/read.js#L19
```js
          var fpath = `${dir}/${entry.name}`
```
(Note how they hardcode the forward slash above, instead of using something like `path.join(dir, entry.name)` or  `path.sep`.)

Windows also understands forward slashes, so this is not to say that `recursive-fs` is necessarily broken on Windows, just not canonical and potentially mixing forward slashes with backslashes in a path string, which may be more of a matter of aesthetics than correctness. The current implementation in the balena CLI uses backslashes on Windows as the native / canonical form, e.g. the result of [path.normalize()](https://nodejs.org/docs/latest-v10.x/api/path.html#path_path_normalize_path). We could call `path.normalize()` on the paths in `filteredFilePaths`, or we could accept that Windows paths will have a mix of forward slashes and backslashes and amend the test cases instead.

Reply:
```
         "[Warn] -----------------------------------------------------------------------------------------------------------------------------------------------------------------"
         "[Warn] The following .dockerignore file(s) will not be used:"
      -  "[Warn] * C:\\var\\lib\\concourse\\worker\\volumes\\live\\0ac9c696-c6d6-4eba-5984-b6b9c8547f60\\volume\\tests\\test-data\\projects\\no-docker-compose\\dockerignore2/lib/.dockerignore"
      +  "[Warn] * C:\\var\\lib\\concourse\\worker\\volumes\\live\\0ac9c696-c6d6-4eba-5984-b6b9c8547f60\\volume\\tests\\test-data\\projects\\no-docker-compose\\dockerignore2\\lib\\.dockerignore"
         "[Warn] When --multi-dockerignore (-m) is used, only .dockerignore files at the"
         "[Warn] root of each service's build context (in a microservices/multicontainer"
         "[Warn] fleet), plus a .dockerignore file at the overall project root, are used."
         "[Warn] See \"balena help push\" for more details."
         "[Warn] -----------------------------------------------------------------------------------------------------------------------------------------------------------------"
```

To be honest it doesn't feel right to display warning messages with a mix of forward slashes and backslashes in a path string like `no-docker-compose\\dockerignore2/lib/.dockerignore` as shown above, so I would favour a solution where this was avoided, preferably displaying the platform's canonical path format (backslashes on Windows). This could mean converting paths at the point where they are displayed. However, an existing assumption in the codebase is that paths use the platform's canonical representation. Indeed, the documentation of the `filterFilesWithDockerignore` function reads:

```ts
/**
 ...
 * The service directory names/paths must be relative to the project root dir
 * and be "normalized" (path.normalize()) before the call to this function:
 * they should use backslashes on Windows, not contain '.' or '..' segments and
 * not contain multiple consecutive path separators like '//'. Also, relative
 * paths must not start with './' (e.g. 'a/b' instead of './a/b').
 */
export async function filterFilesWithDockerignore(
```

We could also take the view that `recursive-fs` 's approach of hardcoding forward slashes consists of "anti-social behaviour", and we could submit a PR to their repo to use `path.join()` instead.

By the way, just because saying this feels relevant, :-) the rules are different for the contents of tar streams. The tar format always uses forward slashes, never backslashes, regardless of the platform. It's like a platform-independent spec of a file format. Hence calls to `toPosixPath` in code such as:

```
async function newTarDirectory(
	...
		pack.entry(
			{
				name: toPosixPath(fileStats.relPath),
```

Also, and again just because this feels relevant, note that user input on Windows (command line arguments such as `--source`) may contain a mix of forward slashes and backslashes and the CLI currently copes with this by normalising user input. On Linux and macOS, however, backslashes are never understood as path component separators.

Reply:
Regardless of the approach we take here, I've sent a [PR to `recursive-fs`](https://github.com/simov/recursive-fs/pull/5) to adjust the paths generated on Windows.

Reply:
For now, I'd like to push forward and use `path.normalize` on the list of filtered files.